btree/btree.c

```diff
@@ -3460,21 +3535,233 @@ static int64_t btree_get_tick(struct m0_sm_op *smop)
 	};
 }
 
+/** Iterator state machine. */
 int64_t btree_iter_tick(struct m0_sm_op *smop)
 {
-	struct m0_btree_op 	*bop = M0_AMB(bop, smop, bo_op);
-	//ToDo: Implement complete destroy tick function.
-	switch(bop->bo_op.o_sm.sm_state)
-	{
-		case P_INIT:
-			return P_ACT;
+	struct m0_btree_op    *bop   = M0_AMB(bop, smop, bo_op);
+	struct td             *tree  = bop->bo_arbor->t_desc;
+	struct m0_btree_oimpl *oi    = bop->bo_i;
+	struct level          *lev;
 
-		case P_ACT:
-			return P_DONE;
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if ((bop->bo_flags & BOF_COOKIE) &&
+		    cookie_is_set(&bop->bo_rec.r_key.k_cookie))
+			return P_COOKIE;
+		else
+			return P_SETUP;
+	case P_COOKIE:
+		if (cookie_is_valid(tree, &bop->bo_rec.r_key.k_cookie))
+			return P_LOCK;
+		else
+			return P_SETUP;
+	case P_SETUP:
+		bop->bo_arbor->t_height = tree->t_height;
+		oi = level_alloc(tree->t_height);
+		if (oi == NULL)
+			return fail(bop, M0_ERR(-ENOMEM));
+		return P_LOCKALL;
+	case P_LOCKALL:
+		if (bop->bo_flags & BOF_LOCKALL)
+			return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN);
+		/** Fall through if LOCKALL flag is not set. */
+	case P_DOWN:
+		oi->i_used  = 0;
+		oi->i_pivot = -1;
+		return node_get(&oi->i_nop, tree, &tree->t_root->n_addr,
+				P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    s = {};
+			struct segaddr child;
 
-		default:
-			return 0;
+			lev = &oi->i_level[oi->i_used];
+			lev->l_node = oi->i_nop.no_node;
+			s.s_node = oi->i_nop.no_node;
+			lev->l_seq = lev->l_node->n_seq;
+			oi->i_key_found = node_find(&s, &bop->bo_rec.r_key);
+			lev->l_idx = s.s_idx;
+
+			if (node_level(s.s_node) > 0) {
+				if (oi->i_key_found) {
+					s.s_idx++;
+					lev->l_idx++;
+				}
+				/**
+				 * If sibling flag is set, check if the node
+				 * has valid left or right index based on next
+				 * previous flag. If valid left/right index
+				 * found. mark this level as pivot level.
+				 */
+				if (((bop->bo_flags & BOF_NEXT) &&
+				    (lev->l_idx < node_count(lev->l_node))) ||
+				    ((bop->bo_flags & BOF_PREV) &&
+				    (lev->l_idx > 0)))
+					oi->i_pivot = oi->i_used;
+
+				node_child(&s, &child);
+				if (!address_in_segment(child)) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, M0_ERR(-EFAULT));
+				}
+				oi->i_used++;
+				return node_get(&oi->i_nop, tree,
+						&child, P_NEXTDOWN);
+			} else	{
+				/* Get sibling index based on PREV/NEXT flag. */
+				lev->l_idx = sibling_index_get(s.s_idx,
+							       bop->bo_flags,
+							       oi->i_key_found);
+				/**
+				 * In the following cases jump to LOCK state:
+				 * 1. Normal pass without sibling load flag.
+				 * 2. Sibling load flag and found record idx
+				 * is less than number of records in leaf node.
+				 * 3. It is the rightmost record in the btree.
+				 */
+				if (!(bop->bo_flags & BOF_SIBLING) ||
+				    (index_is_valid(lev)) ||
+				    (oi->i_pivot == -1))
```

Comment:
`(index_is_valid(lev)` check is for the case where we are traversing down the tree with `BOF_SIBLING`. And we find that the leaf node has changed so that the search key's next/previous index is now valid. In this case we do not have to load sibling node, we can directly jump to return the nex/prev key record.
`oi->i_pivot == -1` check is to detect that we are at the boundary keys of the btree.